### ISSUE TITLE
Request: Gif button

### ISSUE THREAD

**Issue by kartelhorny** [futo-org/android-keyboard#293]
i think it'd be neat to be able to locally store your favourite gifs by saving them in a dedicated place or add an embed option for copied links in your clipboard

**Comment by anseltaft** [comment]
Rad idea since the keyboard/app would start to slurp up heaps of data if it pulled gifs from the web.

A future problem to note: once you have a few hundred it would be increasingly hard to locate the one you want without adding some keywords to aid in location.

**Comment by ceneblock** [comment]
> Rad idea since the keyboard/app would start to slurp up heaps of data if it pulled gifs from the web.
> 
> A future problem to note: once you have a few hundred it would be increasingly hard to locate the one you want without adding some keywords to aid in location.

Could a third party service be used?

Maybe a red button ( Something to indicate that it is going to the internet.) for giphy or tenor? A blue one for local gifs and a purple for hybrid??

 Once a gif is selected, it's cached along with its related [hash]tags to make search easier.

I don't know if those sites would go with that idea though..

**Comment by MulverineX** [comment]
Duplicate of #83

**Comment by abb128** [comment]
Gif search is primarily used in messaging apps like Discord, Telegram, Signal. Practically all of them seem to include a gif search feature of their own.

It's technically possible to implement online gif search while keeping the main keyboard offline by making a separate online extension app that exposes a ContentProvider for the keyboard to use, but ultimately we then have to pick a specific gif search provider and all of them seem to collect your data for targeted advertising. I'm also not a super big fan of adding a button to the keyboard that uploads your search query to somewhere if we can avoid it, and this seems like one of those avoidable cases considering there's already another gif button in front of the user most of the time.

If there are any apps you use that you really need gif search for and they don't have a builtin gif search, leave a comment, but I'm closing this as unplanned for now.

--

Regarding offline or local gifs, I'm not 100% sure how this could be designed in a way that isn't overly specialized or sucks to use. Maybe once we support images in clipboard manager you'll be able to add your gifs there

**Comment by MulverineX** [comment]
A UI that's consistent across all messaging apps and uses the same external GIF search is a very good thing for UX. Presently, you have a very inconsistent experience across apps which is a giant pain. Also, in some apps there's key differences between how attachments behave vs GIFs that are "native" to the app. The former is often more versatile. Finally, if additional features like GIF favoriting and URL imports to that were added, it would elevate the experience across all apps, and would be a more convenient way of using the same GIFs across platforms. 

Again, no one has to be forced to use this, privacy concerns should be relegated to the user's decision.

**Comment by MulverineX** [comment]
Claiming that a keyboard-based GIF selector isn't needed is like claiming that an emoji picker isn't needed.

**Comment by Happ1ness-dev** [comment]
Really unfortunate. I was hoping gif search would be added eventually, so I can finally switch from gboard. This is literally the last thing holding me back right now.

**Comment by ceneblock** [comment]
I certainly understand the privacy concerns and it appears to be a point defeating request.

Can it be an "opt-in" setting? This way it is still available, but by default there it is a completely offline and privacy respecting keyboard.

**Comment by ratnik2** [comment]
I use matrix - element X android and it does not have GIF support in its composer. Picking GIFs from the keyboard would allow me to use them on element and when I use Instagram on the web. Also selecting GIFs from one place (FUTO keyboard) when I message people is a much better and intuitive design (in my opinion) than using the builtin composers in apps.

**Comment by HadManySons** [comment]
New to FUTO Keyboard. Any update on this? Or somewhere else it is being tracked?

**Comment by Zvonimir-FUTO** [comment]
> New to FUTO Keyboard. Any update on this? Or somewhere else it is being tracked?

Please read https://github.com/futo-org/android-keyboard/issues/293#issuecomment-3024858791

**Comment by HealingSouls** [comment]
> Gif search is primarily used in messaging apps like Discord, Telegram, Signal. Practically all of them seem to include a gif search feature of their own.
> 
> It's technically possible to implement online gif search while keeping the main keyboard offline by making a separate online extension app that exposes a ContentProvider for the keyboard to use, but ultimately we then have to pick a specific gif search provider and all of them seem to collect your data for targeted advertising. I'm also not a super big fan of adding a button to the keyboard that uploads your search query to somewhere if we can avoid it, and this seems like one of those avoidable cases considering there's already another gif button in front of the user most of the time.
> 
> If there are any apps you use that you really need gif search for and they don't have a builtin gif search, leave a comment, but I'm closing this as unplanned for now.
> 
> --
> 
> Regarding offline or local gifs, I'm not 100% sure how this could be designed in a way that isn't overly specialized or sucks to use. Maybe once we support images in clipboard manager you'll be able to add your gifs there

Very disappointed in this decision.     I use GIFs virtually every day in text and without them this keyboard becomes useless.  I have no desire to have to move back to gboard to put in a gif and then switch back multiple times a day.
The other apps you mentioned are closed systems, so sure I can use a gif in signal, to send to the 5 people in my life that use signal, but can't to the 99% of people in my life who don't use signal.

**Comment by HealingSouls** [comment]
> A UI that's consistent across all messaging apps and uses the same external GIF search is a very good thing for UX. Presently, you have a very inconsistent experience across apps which is a giant pain. Also, in some apps there's key differences between how attachments behave vs GIFs that are "native" to the app. The former is often more versatile. Finally, if additional features like GIF favoriting and URL imports to that were added, it would elevate the experience across all apps, and would be a more convenient way of using the same GIFs across platforms.
> 
> Again, no one has to be forced to use this, privacy concerns should be relegated to the user's decision.

Correct.  I'm using FUTO because I wanted better voice to text, not because I don't want my keyboard connecting to the internet, although that is comforting.  But not if it means a daily use feature is missing and I have to go back to gboard to use what FUTO is missing.

**Comment by happybuny005** [comment]
> > Gif search is primarily used in messaging apps like Discord, Telegram, Signal. Practically all of them seem to include a gif search feature of their own.
> > It's technically possible to implement online gif search while keeping the main keyboard offline by making a separate online extension app that exposes a ContentProvider for the keyboard to use, but ultimately we then have to pick a specific gif search provider and all of them seem to collect your data for targeted advertising. I'm also not a super big fan of adding a button to the keyboard that uploads your search query to somewhere if we can avoid it, and this seems like one of those avoidable cases considering there's already another gif button in front of the user most of the time.
> > If there are any apps you use that you really need gif search for and they don't have a builtin gif search, leave a comment, but I'm closing this as unplanned for now.
> > --
> > Regarding offline or local gifs, I'm not 100% sure how this could be designed in a way that isn't overly specialized or sucks to use. Maybe once we support images in clipboard manager you'll be able to add your gifs there
> 
> Very disappointed in this decision. I use GIFs virtually every day in text and without them this keyboard becomes useless. I have no desire to have to move back to gboard to put in a gif and then switch back multiple times a day. The other apps you mentioned are closed systems, so sure I can use a gif in signal, to send to the 5 people in my life that use signal, but can't to the 99% of people in my life who don't use signal.

This [app](https://f-droid.org/packages/com.gifboard/) is pretty good when paired with Futo. It's open source. The only dicy thing about it is it connects to Google search to get the gifs.

**Comment by MulverineX** [comment]
at some point you have to connect to a service like that to get the GIF. also, the keyboard layout of the search on that keyboard is different from FUTO Keyboard, which is a massive pain.
